knative-operator/pkg/common/util.go

```diff
@@ -34,14 +34,30 @@ func IngressNamespace(servingNamespace string) string {
 	return servingNamespace + "-ingress"
 }
 
-// buildImageOverrideMapFromEnviron creates a map to overrides registry images
-func buildImageOverrideMapFromEnviron() map[string]string {
+// BuildImageOverrideMapFromEnviron creates a map to overrides registry images
+func BuildImageOverrideMapFromEnviron(environ []string) map[string]string {
 	overrideMap := map[string]string{}
 
-	for _, e := range os.Environ() {
+	for _, e := range environ {
 		pair := strings.SplitN(e, "=", 2)
 		if strings.HasPrefix(pair[0], "IMAGE_") {
-			name := strings.SplitN(pair[0], "_", 2)[1]
+			suffix := strings.SplitN(pair[0], "_", 2)[1]
+
+			var name string
+
+			// convert
+			// "IMAGE_foo=docker.io/foo"
+			// "IMAGE_eventing-controller_eventing-controller=docker.io/foo2"
+			// to
+			// fo0: docker.io/foo
+			// eventing-controller/eventing-controller: docker.io/foo2
+			suffixParts := strings.SplitN(suffix, "_", 2)
+			if len(suffixParts) == 2 {
+				name = fmt.Sprintf("%s/%s", suffixParts[0], suffixParts[1])
+			} else {
+				name = fmt.Sprintf("%s", suffix)
+			}
+
```

Comment:
How about something along these lines? To me using a negative for the count is more clear -- and we really should consider more than 2 underscores an error, since they're illegal in container names, I think. The comments aren't especially enlightening because you use the same name for both deployment and container -- I can't tell which is which! :)
```suggestion
			// convert
			// "IMAGE_container=docker.io/foo"
			// "IMAGE_deployment_container=docker.io/foo2"
			// to
			// container: docker.io/foo
			// deployment/container: docker.io/foo2
			var name string
			parts := strings.SplitN(pair[0], "_", -1)
			if len(parts) == 3 {
				name = fmt.Sprintf("%s/%s", parts[1], parts[2])
			} else {
				name = parts[1]
			}

```